### ISSUE TITLE
linux4.10 breaks LUKS on non-AES-NI machines

### ISSUE THREAD

**Issue by ibabushkin** [voidlinux/void-packages#5871]
The update to kernel `linux4.10_1` gives the following error on boot (or any cryptsetup invocation on LUKS volumes encrypted with `aes-xts-plain64`):

```
device-mapper: table: 254:8: crypt: Error allocating crypto tfm
Check that kernel supports eas-xts-plain64 cipher
```

A few users on IRC confirmed this. Turns out, none of them had AES-NI support on their CPUs (myself included).

Is anyone aware of any changes to the kernel and/or config that could be leading to this?
```
void-packages/srcpkgs/ $ diff linux4.9/files/x86_64-dotconfig linux4.10/files/x86_64-dotconfig
``` 
doesn't yield anything overly suspicious to me.

EDIT:
here's someone else's `/proc/crypto`, obtained from the dracut rescue shell:
https://gist.github.com/ibabushkin/a475eaa901cba5636118c7b4467948c2

**Comment by MrDetonia** [comment]
This is fixed by adding the `xts` kernel module to your initramfs. It seems there is some dependency issue in the current kernel version.

`/etc/dracut.conf.d/drivers.conf`:
```
add_drivers+="xts"
```

Thanks to people in IRC for help debugging.

**Comment by ibabushkin** [comment]
I can confirm.

**Comment by MrDetonia** [comment]
Should it be Dracut's job to include `xts` in the initramfs automatically? I'm not sure whether that should be assumed when using the `crypt` module or not; seeing as people may be using other ciphers.

If it already should be doing this, it seems 4.10 breaks this behaviour for some reason.

**Comment by ibabushkin** [comment]
On 2017-03-03, Zac wrote:
> Should it be Dracut's job to include `xts` in the initramfs automatically? I'm not sure whether that should be assumed when using the `crypt` module or not; seeing as people may be using other ciphers.
> 
> If it already should be doing this, it seems 4.10 breaks this behaviour for some reason.
That's a good question. The kernel mailing list seems to mention
the regression indepently of any initramfs generator, but I guess
it would be better to check, how exactly the same functionality
was realized under 4.9 and earlier. If possible, reproduce that,
as it was pretty much "works out of the box".

**Comment by chneukirchen** [comment]
Patch is at https://www.spinics.net/lists/stable/msg160678.html

I propose to add this if its not in 4.10.2 anyway.

**Comment by ibabushkin** [comment]
The Update to 4.10.2 includes the given patch, generating a dracut image without explicit module directives includes both `ecb` and `xts`, booting works. If someone else can confirm, I guess it's safe to close.

**Comment by MrDetonia** [comment]
That works for me, booting up fine without anything extra in my dracut config.

**Comment by ibabushkin** [comment]
Alright, closing this now.

**Comment by sunaku** [comment]
@MrDetonia are you sure there are no remnants of your `/etc/dracut.conf.d/drivers.conf` snippet still active in your system? :open_mouth: My system still hits this issue _unless_ I add that snippet to my dracut config.

**Comment by ibabushkin** [comment]
@sunaku on my system without AES-NI, this works correctly without any extra configuration with kernel 4.10.13_1. I haven't tested with 4.10.14_1, but this *should* not be an issue. Are you sure it's the same issue and not just the same behaviour?

**Comment by sunaku** [comment]
Yes, I just reproduced this very issue on [a fresh install of VoidLinux](https://wiki.voidlinux.eu/Install_LVM_LUKS) on another machine using the `void-live-x86_64-20170220.iso ` image, which comes with Linux 4.9 by default. :hammer_and_wrench:  Under that default 4.9 kernel, my LUKS passphrase is correctly accepted at boot. :joy: However, by simply upgrading to the latest 4.10 kernel, I encountered the same issue described in the original bug report at the top of this page. :scream:
